Edit team page for Instructors only.

muzaffaryousaf opened the pull request:
This PR adds the Edit teams page for instructors.
For this we show the Edit team button on Team's profile page. 
It should be visible for instructors(global staff, course staff & privileged users) only for now.

**Jira tickets:**
* https://openedx.atlassian.net/browse/TNL-1914
* https://openedx.atlassian.net/browse/TNL-3045

Sandbox: http://muzaffar-team-edit.sandbox.edx.org/

muzaffaryousaf commented:
jenkins run js

muzaffaryousaf commented:
@symbolist @marcotuts @catong Can you please give it a look. The sandbox is available [here](http://muzaffar-team-edit.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/teams)
Thanks

marcotuts commented on lms/static/sass/elements/_system-feedback.scss:
```diff
@@ -145,6 +145,13 @@
   }
 }
 
+.wrapper-msg {
```
It wasn't clear to me what the goal of having this abstraction here for the fixed variant of the wrapper. Should the .wrapper-msg-fixed class also be including other styles to set the fixed behavior?

muzaffaryousaf replied:
@marcotuts I wanted to have similar message banner that stays on the UI all the time and that looks similar to the existing one. 

Should we add fixed postfix with other rules as well to avoid the errors and make same copy of new rules but with different selectors that includes fixed ???

marcotuts replied:
To clarify, it seemed like .wrapper-msg-fixed and .wrapper-msg are both the same currently, since they both just extend wrapper-msg. Is that correct?

muzaffaryousaf replied:
yes, we hide/show the banner by wrapper-msg, that's why i named it different.

muzaffaryousaf replied:
@marcotuts Are we good here ?

marcotuts replied:
+1

symbolist commented on lms/djangoapps/teams/views.py:
```diff
@@ -116,6 +117,7 @@ def get(self, request, course_id):
             "topics_url": reverse('topics_list', request=request),
             "teams_url": reverse('teams_list', request=request),
             "team_memberships_url": reverse('team_membership_list', request=request),
+            "teams_detail_url": reverse('teams_detail', args=['team_id']),
```
Nit: Can we move this a line up?

marcotuts commented on lms/static/sass/elements/_controls.scss:
```diff
@@ -404,7 +404,7 @@
   }
 
-  .btn-link {
+  .btn-secondary {
```
This is something we added in an earlier teams story that I haven't had a chance to review / cleanup, but now is a good opportunity I believe. We currently use this in the teams details sidebar for "Leave Team" if I'm not mistaken. 

Instead of using the btn-secondary name here, which is used already for other pattern library button styles above, we should update this btn-link styling to echo the UX button style that has no padding to the left / right.

This is a link to the "btn-link" pattern library: https://github.com/edx/ux-pattern-library/blob/master/_src/pattern-library/sass/components/_buttons.scss#L229

![image](https://cloud.githubusercontent.com/assets/2023680/9411499/3a5bd936-47f4-11e5-849b-8644bf1896a8.png)

muzaffaryousaf replied:
@marcotuts I will add btn-link class to Leave Team button. Should i use the same `btn-secondary` for edit team button or do we have any existing rule for that?

marcotuts replied:
Thanks for doing that. (btn-link --> Leave Team)

I would extend %btn-pl-secondary-base (which is in this _controls.scss file) for Edit Team, as that button can have padding / spacing around it.  That extend should map onto our secondary styling as shown in the image below.

![image](https://cloud.githubusercontent.com/assets/2023680/9450504/ed5f88c8-4a76-11e5-9358-cab328a6267b.png)

symbolist commented on lms/djangoapps/teams/static/teams/js/views/edit_team.js:
```diff
@@ -28,11 +28,18 @@
-                    this.teamModel.url = this.teamsUrl;
+                    if (this.action === 'create') {
+                        this.teamModel = new TeamModel({});
+                        this.teamModel.url = this.teamsUrl;
```
Can this new url not be passed when initializing in the previous line?

symbolist commented on lms/djangoapps/teams/static/teams/js/views/edit_team.js:
```diff
@@ -28,11 +28,18 @@
                     this.languages = options.teamParams.languages;
                     this.countries = options.teamParams.countries;
                     this.primaryButtonTitle = options.primaryButtonTitle || 'Submit';
```
If we are passing in the action, I don't really see why this needs to be passed in via options. Just set it in the conditional below.

symbolist commented on lms/djangoapps/teams/static/teams/js/views/edit_team.js:
```diff
@@ -74,7 +81,11 @@
-                    this.$el.html(_.template(editTeamTemplate)({primaryButtonTitle: this.primaryButtonTitle}));
+                    this.$el.html(_.template(editTeamTemplate)({
+                        primaryButtonTitle: this.primaryButtonTitle,
+                        isNew: this.action === 'create',
```
Passing in the `action` instead of a boolean will be nicer since it describes the purpose of the form.

symbolist commented on lms/djangoapps/teams/static/teams/js/views/edit_team.js:
```diff
@@ -91,20 +102,28 @@
                 },
 
-                createTeam: function (event) {
+                createUpdateTeam: function (event) {
```
Can we go for `createOrUpdate` team?

symbolist commented on lms/djangoapps/teams/static/teams/js/views/edit_team.js:
```diff
@@ -91,20 +102,28 @@
+                            language: _.isNull(teamLanguage) ? '' : teamLanguage,
+                            country: _.isNull(teamCountry) ? '' : teamCountry
+                        },
+                        defaultParams = {
```
I think `saveOptions` would be a more helpful name.

symbolist commented on lms/djangoapps/teams/static/teams/js/views/edit_team.js:
```diff
@@ -186,8 +205,15 @@
+                    var url;
+                    if (this.action === 'create') {
+                        url = 'topics/' + this.topicID;
+                    } else {
```
In case of non-bools, specifying the options explicitly ("`else if (this.action === 'edit')`) makes the code easier to understand.

symbolist replied:
You can also do this for the other conditionals.

symbolist commented on lms/djangoapps/teams/static/teams/js/views/edit_team.js:
```diff
@@ -186,8 +205,15 @@
 
-                goBackToTopic: function () {
-                    Backbone.history.navigate('topics/' + this.topicID, {trigger: true});
+                cancelTeam: function (event) {
```
How about `cancelAndGoBack`?

symbolist commented on lms/djangoapps/teams/static/teams/js/views/edit_team.js:
```diff
@@ -74,7 +81,11 @@
 
                 render: function() {
-                    this.$el.html(_.template(editTeamTemplate)({primaryButtonTitle: this.primaryButtonTitle}));
+                    this.$el.html(_.template(editTeamTemplate)({
```
Nit: needs a space after `)`.

symbolist commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -219,6 +222,41 @@
+                                main: new TeamEditView({
+                                    action: 'edit',
+                                    teamEvents: self.teamEvents,
+                                    tagName: 'create-new-team',
```
If this is going to be the same in both cases, we don't need to make it an option.

symbolist commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -303,14 +342,17 @@
                             });
-                            var teamJoinView = new TeamJoinView({
+
+                            var TeamProfileActionsView = new TeamProfileHeaderActionsView({
```
Should be `teamProfileActionsView `.

muzaffaryousaf replied:
We use `TeamProfileActionsView ` convention for views, models and collections names.

symbolist commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -219,6 +222,41 @@
+                 */
+                editTeam: function (topicID, teamID) {
+                    var self = this;
+                    this.getTeamsView(topicID).done(function (teamsView) {
```
This is only being used to get the `teamParams`. Instead we should create a function `get_team_params(topicId)` and call it wherever the context is required.

cahrens replied:
Is this function still relevant?

catong commented on lms/djangoapps/teams/static/teams/js/spec/views/edit_team_spec.js:
```diff
@@ -31,33 +38,57 @@ define([
                 var message = teamEditView.$('.wrapper-msg');
                 expect(message.hasClass('is-hidden')).toBeFalsy();
-                expect(message.find('.title').text().trim()).toBe("Your team could not be created!");
+                var actionMessage = teamAction === 'create' ? 'Your team could not be created!' : 'Your team could not be updated!';
```
Are these visible in the UI? If yes, sorry if I missed this in some earlier story. We should change the exclamation marks to periods: "Your team could not be created." "Your team could not be updated."

muzaffaryousaf replied:
done.

catong commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -219,6 +222,41 @@
+                            self.mainView = new ViewWithHeader({
+                                header: new HeaderView({
+                                    model: new TeamsHeaderModel({
+                                        description: gettext("You can edit the details of a team here, and should separately let your team members know about changes."),
```
"If you edit any team details, you should let the team's members know about the changes."

muzaffaryousaf replied:
Done.

catong commented on lms/djangoapps/teams/static/teams/templates/edit-team.underscore:
```diff
@@ -2,7 +2,13 @@
-            <h3 class="title"><%- gettext("Your team could not be created!") %></h3>
+            <h3 class="title">
+                <% if (isNew) { %>
+                    <%- gettext("Your team could not be created!") %>
```
Please replace the exclamation marks in these strings with periods. (Line 7 and line 9)
"Your team could not be created." "Your team could not be updated."

muzaffaryousaf replied:
done.

catong commented on lms/djangoapps/teams/static/teams/templates/edit-team.underscore:
```diff
@@ -11,10 +17,31 @@
+                <p>
+                    <%=
+                        interpolate_text(
+                            _.escape(gettext("The team that you are editing has {total_members} members. Communicate important changes to students prior to updating the team when possible.")),
```
It seems a bit strange to have information about the member count and an advisory message combined in this message. 
I'm wondering if we can do without the second sentence, which we can cover in the text under the "Edit Team" heading. Would it make sense here to just mention the number of members?
Or, if we want to remind staff about notifying team members of changes here, is there another logical place to put the count of members?
@marcotuts @muzaffaryousaf And @explorerleslie (when you get back), what are your thoughts?

catong replied:
@muzaffaryousaf Here's my suggestion:
""The team that you are editing has {total_members} members. Notify team members prior to making significant changes."

catong commented on lms/djangoapps/teams/static/teams/templates/edit-team.underscore:
```diff
@@ -38,7 +65,7 @@
         <%=
             interpolate_text(
-                _.escape(gettext("{primaryButtonTitle} {span_start}a new team{span_end}")),
+                _.escape(gettext("{primaryButtonTitle} {span_start}a team{span_end}")),
```
@muzaffaryousaf I can't tell where this appears or how it is meant to read.

muzaffaryousaf replied:
@catong This is sr for Create/Update button on Create/Update team page.

catong replied:
Thanks, @muzaffaryousaf, no changes here.

muzaffaryousaf commented:
@lou-wang Can you please give it a look from product perspective. This PR adds the Edit team functionality for user i.e. course staff, global staff and/or Moderator/Community TA/Administrator. Thanks.

Here are the links to ticket and sandbox:
**Jira tickets:**
* https://openedx.atlassian.net/browse/TNL-1914
* https://openedx.atlassian.net/browse/TNL-3045

Sandbox: http://muzaffar-team-edit.sandbox.edx.org/

muzaffaryousaf commented:
@lou-wang FYI, you can navigate to the team edit page from [here](http://muzaffar-team-edit.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/teams/#teams/0/first-team-by-muzaffar) and the edit team button should be visible for only those users with above mentioned roles. Thanks.

muzaffaryousaf commented:
@catong I've addressed your comments, is there any thing else that you want me to change ? Thanks.

catong commented on lms/djangoapps/teams/static/teams/templates/edit-team.underscore:
```diff
@@ -48,7 +75,7 @@
         <%=
             interpolate_text(
-                _.escape(gettext("Cancel {span_start}a new team{span_end}")),
+                _.escape(gettext("Cancel {span_start} team {span_end}")),
```
@muzaffaryousaf I think we need a verb here to make it clear what action is being cancelled. If we can make the distinction between "create" and "update" can we make it "Cancel creating team" and "Cancel updating team"? If we can't make the distinction, then "Cancel creating or updating team"? @clrux any thoughts?

cptvitamin replied:
@muzaffaryousaf @catong I'm not a huge fan of Cancel buttons to begin with. Is it absolutely necessary in this case?  Not hitting submit (create/update) is basically the same as Cancel.  I find that having cancel buttons only result in users mistakenly hitting the cancel button and swearing at their computers for having lost form data they worked hard to enter.  I'm in favor of just omitting it all together unless it is necessary for some reason.

muzaffaryousaf replied:
@cptvitamin It is product requirement and already implemented for Create New team page. Maybe @lou-wang or @explorerleslie @marcotuts  can confirm ^^.

cptvitamin replied:
If its a product requirement, perhaps we should carefully re-consider it.  Jakob Nielson questioned the efficacy of the Cancel button back in 2000:
http://www.nngroup.com/articles/reset-and-cancel-buttons/

We have to ask ourselves if its necessary.  99 times out of 100 I find it is not.

marcotuts replied:
Nice reference there @cptvitamin. :) Throwback for sure, but still relevant. 

In this case we'd like to keep the cancel buttons, in part because it extends Studio's existing patterns, but also because it falls into the sweet spot of usage where cancel buttons actually are #NeilsonApproved. 

While many web applications use pages primarily, from time to time confirmation modals / messages are used to double check actions / operations that might be destructive or accidentally initiated. Neilson describes it this way under the "Cancel: Use sparingly" section. 

"Offer a Cancel button when users may fear that they have committed to something they want to avoid. Having an explicit way to Cancel provides an extra feeling of safety that is not afforded by simply leaving."

cptvitamin replied:
great @marcotuts.  If that condition applies here, then lets not cancel the cancel button ;)  I don't have any background on this control, I was just reviewing code.  If we are going to keep the cancel button, I prefer to add some contextual screen reader text that clarifies exactly what it is that is being cancelled as @catong originally suggested.

marcotuts replied:
Ah ok, yeah the screenreader text suggested makes sense to me. (cc'ed @catong )

(Also, we can revisit our use of this semi-modal pattern moving forward, as it will increasingly impact more pages and part of the experience in future quarters. (ex. Create New Discussion Topic / Post, etc.)

symbolist commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -564,78 +564,116 @@ def test_navigation_links(self):
-        self.verify_page_header()
+        self.verify_page_header(
+            title='Create a New Team',
+            description='Create a new team if you can\'t find existing teams to '
```
These strings can be made constants at the top of the file.

symbolist replied:
Though if they are used only here, it is fine to leave them like this.

symbolist commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -747,6 +788,133 @@ def test_user_can_cancel_the_team_creation(self):
+    def test_discussion_privileged_user_can_see_edit_button(self, role):
+        """
+        Scenario: The user with specified role should see the edit team button.
+        Given I am user with role for a course with a team
```
"with privileged role"

symbolist commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -747,6 +788,133 @@ def test_user_can_cancel_the_team_creation(self):
+
+        self.team_page.visit()
+        self.teams_page.wait_for_page()
+        self.assertTrue(self.team_page.edit_team_button_present)
```
Might as well also check that the team can be successfully edited.

muzaffaryousaf replied:
Done.

symbolist commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -564,78 +564,116 @@ def test_navigation_links(self):
+        self.create_or_edit_team_page.value_for_dropdown_field(field_id='country', value='Pakistan')
 
-    def test_user_can_see_create_team_page(self):
+    def verify_create_or_edit_page(self):
```
`verify_team_data` would be more helpful.

symbolist commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -564,78 +564,116 @@ def test_navigation_links(self):
-        self.set_team_configuration({'course_id': self.course_id, 'max_team_size': 10, 'topics': [self.topic]})
+        self.set_team_configuration(
+            {'course_id': self.course_id, 'max_team_size': 10, 'topics': [self.topic]},
+            global_staff=True
```
Instead of using a global staff user for all the tests, it would be better to use a course staff user.

symbolist commented on lms/djangoapps/teams/static/teams/js/spec/views/team_profile_header_actions_spec.js:
```diff
@@ -0,0 +1,254 @@
+                expect(view.$('.action-edit-team').length).toEqual(showEditButton ? 1 : 0);
+            };
+
+            it('can render itself for staff or privileged user', function () {
```
How about "renders for staff or privileged user" and "does not render for non-staff and non-privileged user"?

symbolist replied:
On second thoughts `TeamProfileHeaderActionsView` does not know any details about the user so "renders when option showEditButton is true" and "does not render when option showEditButton is false" will be more accurate.

muzaffaryousaf commented on lms/djangoapps/teams/static/teams/js/spec/views/team_profile_header_actions_spec.js:
```diff
@@ -0,0 +1,254 @@
+define([
```
For Reviewers: I've renamed this file and added `EditButton` spec. Also, i've moved some common helper methods to the main spec `TeamProfileHeaderActionsView ` so we can use them in `JoinButton` & `EditButton`

cahrens replied:
Can you move files instead of deleting/adding? That makes it possible to see what actually changed (and keeps file history).

muzaffaryousaf replied:
I've renamed this file with others but i don't know why only this file is different from others. Let me give it a look.

symbolist commented on lms/djangoapps/teams/static/teams/js/spec/views/edit_team_spec.js:
```diff
@@ -4,13 +4,14 @@ define([
+    describe('CreateEditTeam', function() {
         var teamsUrl = '/api/team/v0/teams/',
-            teamsData = {
+            createTeamsData = {
```
Nit: should be singular "Team". Also on line 26.

symbolist commented on lms/djangoapps/teams/static/teams/js/spec/views/edit_team_spec.js:
```diff
@@ -31,33 +38,57 @@ define([
                 var message = teamEditView.$('.wrapper-msg');
                 expect(message.hasClass('is-hidden')).toBeFalsy();
-                expect(message.find('.title').text().trim()).toBe("Your team could not be created!");
+                var actionMessage = teamAction === 'create' ? 'Your team could not be created.' : 'Your team could not be updated.';
```
It will be a bit safer to put the conditional in `()`.

symbolist commented on lms/djangoapps/teams/static/teams/js/spec/views/edit_team_spec.js:
```diff
@@ -67,15 +98,15 @@ define([
         });
 
-        it('can render itself correctly', function () {
+        var assertTeamsRendering = function() {
```
`assertFieldsRenderCorrectly` or `assertFormRendersCorrectly`.

symbolist commented on lms/djangoapps/teams/static/teams/js/spec/views/edit_team_spec.js:
```diff
@@ -83,32 +114,47 @@ define([
+            }
+        };
+
+        var teamsMethod = function() {
```
`requestMethod`?

symbolist commented on lms/djangoapps/teams/static/teams/js/spec/views/edit_team_spec.js:
```diff
@@ -83,32 +114,47 @@ define([
+            expect(Backbone.history.navigate.calls[0].args).toContain(expectedUrl);
+        };
+
+        var assertValidationsWithEmptyFields = function(that) {
```
`assertValidationMessagesShowWhenFieldsEmpty`?

symbolist replied:
or maybe just `assertValidationMessagesWhenFieldsEmpty`?

symbolist commented on lms/djangoapps/teams/static/teams/js/spec/views/edit_team_spec.js:
```diff
@@ -119,13 +165,13 @@ define([
-                teamEditView = createTeamEditView(),
-                teamName = new Array(500 + 1).join( '$'),
-                teamDescription = new Array(500 + 1).join( '$' );
+        var assertValidationsForInvalidData = function(that) {
```
`assertValidationMessagesWhenInvalidData`

explorerleslie commented:
@muzaffaryousaf I'm back and can review this PR instead of @lou-wang by end of day. I'm assuming I'm not blocking merge because @symbolist just left a bunch of comments but please let me know if I'm blocking.

symbolist commented on lms/djangoapps/teams/static/teams/js/spec/views/edit_team_spec.js:
```diff
@@ -141,48 +187,114 @@ define([
-        it("shows an error message for HTTP 500", function () {
-            var teamEditView = createTeamEditView(),
-                requests = AjaxHelpers.requests(this);
+        var assertServerErrors = function(that, teamsData, teamsUrl, errorCode) {
```
`assertShowMessageOnError`?

symbolist commented on lms/djangoapps/teams/static/teams/js/spec/views/edit_team_spec.js:
```diff
@@ -141,48 +187,114 @@ define([
 
-            teamEditView.$('.u-field-name input').val(teamsData.name);
-            teamEditView.$('.u-field-textarea textarea').val(teamsData.description);
+        var assertExpectedUrl = function(expectedUrl) {
```
`assertRedirectsToCorrectUrlOnCancel`?

symbolist commented:
@muzaffaryousaf I am done with a pass. Let me know when you need me to look at this again.

muzaffaryousaf commented:
@symbolist I'm done with changes, please have look again. Thanks.

muzaffaryousaf commented:
@explorerleslie I"m done with changes that usman suggested. You can look this w.r.t product perspective.

muzaffaryousaf commented:
@marcotuts @catong Would you mind to have final look. I've addressed your changes. Thanks.

symbolist commented:
@cahrens This is ready for a review from the Cambridge team. Thanks.

catong commented on lms/djangoapps/teams/static/teams/templates/edit-team.underscore:
```diff
@@ -11,10 +17,31 @@
+                <p>
+                    <%=
+                        interpolate_text(
+                            _.escape(gettext("The team that you are editing has {total_members} members. Notify team members prior to making significant changes.")),
```
@muzaffaryousaf I just noticed that this always says "members" even when the team member count is 1 ("...has 1 members."). Is there a way to make the noun agree with the number?

muzaffaryousaf replied:
@catong sure. Will make the change.

catong commented:
@muzaffaryousaf I only had one query about making the "members" vs "member" in the string work with the number of members. Otherwise, looks great.

cahrens commented:
I will take a look at this today.

marcotuts commented:
This looks great, thanks @muzaffaryousaf. I'm going to add a UX story to the backlog for FED cleanup / review here once teams is complete, since I think there might be some further cleanup / reuse we could do here. For now though, this is good from my perspective. :+1:

muzaffaryousaf commented:
**REVIEWERS:**
Sandbox is updated and you can access it [here](http://muzaffar-team-edit.sandbox.edx.org/)

cahrens commented:
I got this error in the JS console when updating a team on the sandbox:

Uncaught TypeError: Cannot read property 'image_url_medium' of undefined(anonymous function) @ teams_tab_factory.d9e9c8269cb8.js?raw:6w.each.w.forEach @ lms-base-vendor.bbc6ad2cc6fb.js:4e.View.extend.renderTeamMembers @ teams_tab_factory.d9e9c8269cb8.js?raw:6e.View.extend.render @ teams_tab_factory.d9e9c8269cb8.js?raw:6e.View.extend.render @ teams_tab_factory.d9e9c8269cb8.js?raw:7e.View.extend.render @ teams_tab_factory.d9e9c8269cb8.js?raw:7(anonymous function) @ teams_tab_factory.d9e9c8269cb8.js?raw:7o @ lms-base-vendor.bbc6ad2cc6fb.js:1p.add @ lms-base-vendor.bbc6ad2cc6fb.js:1e.View.extend.browseTeam @ teams_tab_factory.d9e9c8269cb8.js?raw:7(anonymous function) @ lms-base-vendor.bbc6ad2cc6fb.js:6(anonymous function) @ lms-base-vendor.bbc6ad2cc6fb.js:6w.some.w.any @ lms-base-vendor.bbc6ad2cc6fb.js:4h.extend.loadUrl @ lms-base-vendor.bbc6ad2cc6fb.js:6h.extend.navigate @ lms-base-vendor.bbc6ad2cc6fb.js:6(anonymous function) @ teams_tab_factory.d9e9c8269cb8.js?raw:7o @ lms-base-vendor.bbc6ad2cc6fb.js:1p.fireWith @ lms-base-vendor.bbc6ad2cc6fb.js:1w @ lms-base-vendor.bbc6ad2cc6fb.js:3d @ lms-base-vendor.bbc6ad2cc6fb.js:3

cahrens commented:
Why did "Leave Team" become so small?

![image](https://cloud.githubusercontent.com/assets/484484/9473607/b893358c-4b29-11e5-83b1-eeb62b34e698.png)

cahrens commented:
Why are we showing this message when there are no team members?

![image](https://cloud.githubusercontent.com/assets/484484/9473822/ead4eecc-4b2a-11e5-9b16-aabdb7704038.png)

cahrens commented on lms/djangoapps/teams/static/teams/js/views/edit_team.js:
```diff
@@ -13,10 +13,12 @@
                 maxTeamNameLength: 255,
                 maxTeamDescriptionLength: 300,
 
+                tagName: 'create-new-team',
```
How is this tagName used? And should it be specific to creating a team?

muzaffaryousaf replied:
This was being introduced in create team page, i've just moved it to here and looks like this is not being used anywhere. Removing it.

cahrens commented on lms/djangoapps/teams/static/teams/js/views/edit_team.js:
```diff
@@ -27,12 +29,20 @@
+                    if (this.action === 'create') {
+                        this.teamModel = new TeamModel({});
+                        this.teamModel.url = this.teamsUrl;
+                        this.primaryButtonTitle = 'Create';
```
i18n!

muzaffaryousaf replied:
Already in edit_team template.

cahrens replied:
What does that mean? You can't just translate primaryButtonTitle as a variable, as the value may not have been scraped. Can you point me to the line?

muzaffaryousaf replied:
https://github.com/edx/edx-platform/pull/9376/files#diff-d1ab4e2ba56dd0e0be2b27a3e9724a8fR73

muzaffaryousaf replied:
@cahrens That's existing code though. I think we need some tweaks there too.

cahrens replied:
This won't work because you can't call gettext on a variable, since that prevents the actual text represented by the variable from being scraped.

cahrens replied:
Even if this is existing, it needs to be fixed-- it shouldn't have passed previous code reviews. It will only translate if some place in our codebase someone happens to have scraped this same text.

muzaffaryousaf replied:
Agree. Will update this accordingly. Thanks for pointing this out.

cahrens replied:
Since you do have a separate i18n story (https://openedx.atlassian.net/browse/TNL-3125), I would be OK with you listing this issue in that ticket and fixing it as a follow-up PR.

cahrens commented on lms/djangoapps/teams/static/teams/js/views/edit_team.js:
```diff
@@ -27,12 +29,20 @@
+                    } else if(this.action === 'edit' ) {
+                        this.teamModel = options.model;
+                        this.teamModel.url = this.teamsDetailUrl.replace('team_id', options.model.get('id'));
+                        this.primaryButtonTitle = 'Update';
```
i18n

muzaffaryousaf replied:
in template.

cahrens replied:
Same issue-- need to scrape the actual text.

cahrens commented on lms/djangoapps/teams/static/teams/js/views/edit_team.js:
```diff
@@ -91,20 +105,28 @@
+                    if (this.action === 'create') {
+                        data.course_id = this.courseID;
+                        data.topic_id = this.topicID;
+                    } else if(this.action === 'edit' ) {
```
Nit: spacing (after "if").

cahrens commented on lms/djangoapps/teams/static/teams/js/views/edit_team.js:
```diff
@@ -186,8 +208,15 @@
+                    var url;
+                    if (this.action === 'create') {
+                        url = 'topics/' + this.topicID;
+                    } else if(this.action === 'edit' ) {
```
Nit: spacing (after "if").

cahrens commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -219,6 +220,39 @@
+                    var self = this;
+                    this.getTeamsView(topicID).done(function (teamsView) {
+                        self.getTeam(teamID, false).done(function(team) {
+                            self.mainView = new ViewWithHeader({
```
This code is very difficult to read and understand. Perhaps pull some of it out into variables so they can be named? At the very least, put a comment in about what "false" signifies on line 228.

muzaffaryousaf replied:
@cahrens I'm refactoring this to use `createViewWithHeader` method instead.

cahrens commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -219,6 +220,39 @@
+                                        title: gettext("Edit Team"),
+                                        breadcrumbs: [
+                                            {
+                                                title: teamsView.main.teamParams.topicName,
```
This breadcrumb seems wrong-- it takes you back to the listing of all teams in the topic. I would expect to go back to the Team view.

cahrens commented on lms/djangoapps/teams/static/teams/templates/edit-team.underscore:
```diff
@@ -38,7 +70,7 @@
     <button class="action action-primary">
         <%=
             interpolate_text(
-                _.escape(gettext("{primaryButtonTitle} {span_start}a new team{span_end}")),
```
The line below (73) is going to need translator comments-- it would be confusing to a translator.

Also, does it matter if whitespace is put between {span_start} and "a team" (and then the {span_end})? If so, this is going to be very brittle for translation.

cahrens replied:
Here I see the translation of the primaryButtonTitle. This won't work because you can't call gettext on a variable, since that prevents the actual text represented by the variable from being scraped.

cahrens commented on lms/djangoapps/teams/static/teams/templates/edit-team.underscore:
```diff
@@ -48,9 +80,10 @@
                 {
-                    'span_start': '<span class="sr">', 'span_end': '</span>'
+                    'span_start': '<span class="sr">', 'span_end': '</span>',
+                    'action': action === 'create' ? 'creating' : 'updating'
```
i18n ("creating", "updating").

This whole thing is going to very difficult for a translator to deal with. I recommend you split it up into 2 different strings (one for the updating case, one for creating), and also add translator comments. The concatenation approach here with individual words being substituted in won't work.

cahrens commented on lms/djangoapps/teams/static/teams/templates/team-profile.underscore:
```diff
@@ -48,7 +48,7 @@
         <% if (isMember) { %>
             <div class="leave-team">
-                <button class="btn btn-link btn-base btn-secondary leave-team-link"><%- gettext("Leave Team") %></button>
+                <button class="btn btn-link leave-team-link"><%- gettext("Leave Team") %></button>
```
I think this is the change that made "Leave  Team" so tiny.

marcotuts replied:
Both font-weight and font-size rules are being applied from the base <button> element styling here.  I can take a look and propose an update to resolve this.

muzaffaryousaf replied:
Done.

cahrens commented on lms/djangoapps/teams/views.py:
```diff
@@ -71,11 +71,12 @@ def get(self, request, course_id):
 
-        if not CourseEnrollment.is_enrolled(request.user, course.id) and \
-                not has_access(request.user, 'staff', course, course.id):
+        if not CourseEnrollment.is_enrolled(request.user, course.id) and not is_staff:
```
Nit: I would leave the code as it was so that we don't evaluate the has_access call unless the caller is not enrolled in the course.

cahrens commented:
I did not get to reviewing the tests today, but did make it through all the non-test code.

marcotuts commented on lms/static/sass/elements/_controls.scss:
```diff
@@ -416,3 +416,20 @@
+    padding: 1px;
+    background: $transparent;
+    background-image: none;
+
```
adding this to the reset list should resolve the size issue Christina noted.

@extend %t-action3; 
@extend %t-strong;

muzaffaryousaf replied:
Done

marcotuts commented on lms/static/sass/elements/_controls.scss:
```diff
@@ -416,3 +416,20 @@
+    &:focus,
+    &:hover {
+        background-image: none !important;
+        background-color: $transparent !important;
```
I would also add box-shadow: none; here, to remove the large box-shadow added by default here.

muzaffaryousaf replied:
Done

marcotuts commented on lms/static/sass/elements/_system-feedback.scss:
```diff
@@ -4,7 +4,7 @@
 
 // UI : message
-.wrapper-msg {
+%wrapper-msg {
```
after reviewing with Leslie, the suggestion here would be to have margin-top: $baseline instead of margin: $baseline. 

The desired effect is to have the fixed message be full width, minus the 20px of margin on top currently being added.

Screenshot:

![image](https://cloud.githubusercontent.com/assets/2023680/9494778/fb6ffc9e-4bd5-11e5-9c16-d5c43f924e82.png)

explorerleslie commented:
@muzaffaryousaf when an error occurs, the message is displayed twice. It should only display once -- use the first one displayed in the screenshot below which better matches the studio pattern. The bar in this case should also use the error color, which is reddish, not yellow, which is used for warnings. 

![image](https://cloud.githubusercontent.com/assets/5643426/9494674/4502bee2-4bd5-11e5-9982-47d189bdbae7.png)

muzaffaryousaf commented:
@explorerleslie What should be the expected breadcrumb for edit page. I think it should be
`All Topics > TEAM_NAME > ` ?  Does it make sense OR we should use existing pattern which is `All Topics > TOPIC_NAME > `

muzaffaryousaf commented:
@explorerleslie 
>Why are we showing this message when there are no team members?
![image](https://cloud.githubusercontent.com/assets/484484/9473822/ead4eecc-4b2a-11e5-9b16-aabdb7704038.png)

Should we hide this banner in case of no member OR do we need to show some other message.?

muzaffaryousaf commented:
@explorerleslie 
> The bar in this case should also use the error color, which is reddish, not yellow, 

I've fixed this but it was part of create team.

muzaffaryousaf commented:
@cahrens I've fixed the profile image issue and addressed your comment except translation. Please continue the review. Thanks.

muzaffaryousaf commented:
@explorerleslie @cahrens @marcotuts @carol FYI, I have updated the sandbox for this. Thanks

explorerleslie commented:
@muzaffaryousaf 

* When editing team, breadcrumb should be All Topics > Topic Name > Team Name. @marcotuts @catong does that make sense? 
* We don't need to show a message when there are zero members of a team; just hide it. 
* All the warning bars need to be full width and have the margin updates per @marcotuts's comments before, but I don't see this reflected in the sandbox. https://github.com/edx/edx-platform/pull/9376#discussion-diff-37981562

catong commented:
@explorerleslie The breadcrumb trail that you describe makes sense to me.
Question for you -- what do you think of the repetition of the text about notifying team members of changes? We have text under the Edit Team form heading as well as in the banner. 
I don't know if it is a UX pattern requirement to always have text under a form heading. If it is not, perhaps we can do without the text there? (@marco?)

muzaffaryousaf commented:
@cahrens bokchoys are failing due to the breadcrumb change. I just need to change the breadcrumb according to the new pattern in js as well as in bokchoy and then we will be all set.

catong commented:
@marcotuts I discussed this with @explorerleslie, and we think that the informational banner is not really necessary, given that the warning about notifying teams of changes is always visible under the "Edit Team" heading, and also that course staff can see the number of members in a team as they go in to edit it. What would you think about not using the banner at all? If we got rid of the banner, I would tweak the wording under "Edit Team" to use the version we currently use in the banner ("If you make significant changes, make sure you notify members of the team before making these changes.")

cahrens commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -813,6 +859,165 @@ def test_user_can_cancel_the_team_creation(self):
+        self.team = self.create_teams(self.topic, num_teams=1)[0]
+        self.team_page = TeamPage(self.browser, self.course_id, team=self.team)
+        self.team_page.visit()
+        self.team_page.wait_for_page()
```
The visit method ends with "wait_for_page", so delete this line.

cahrens commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -767,16 +809,20 @@ def test_user_can_create_new_team_successfully(self):
         """
+        AutoAuthPage(self.browser, course_id=self.course_id).visit()
+        self.browse_teams_page.visit()
+        self.browse_teams_page.wait_for_page()
```
The visit method ends with "wait_for_page", so delete this line.

cahrens commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -813,6 +859,165 @@ def test_user_can_cancel_the_team_creation(self):
+        """
+        AutoAuthPage(self.browser, course_id=self.course_id).visit()
+        self.team_page.visit()
+        self.teams_page.wait_for_page()
```
The visit method ends with "wait_for_page", so delete this line.

cahrens commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -813,6 +859,165 @@ def test_user_can_cancel_the_team_creation(self):
+        And I click Update button
+        Then I should see the page for my team with updated data
+        """
+        self.verify_form_values_on_team_page(
```
How is this test point already in "edit" mode (for verifying the field forms)?

muzaffaryousaf replied:
Correct, Update the method name. Sorry for confusion.

cahrens commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -629,78 +629,120 @@ def test_navigation_links(self):
+        )
+
+    def verify_form_values_on_team_page(self, name, description, location, language):
+        """Verify the values for create/edit form."""
```
I think this doc string is wrong-- the create/edit form isn't what is being tested.

Can this method just be verify_team or verify_team_values?

cahrens commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -813,6 +859,165 @@ def test_user_can_cancel_the_team_creation(self):
+        self.teams_page.wait_for_page()
+        self.assertTrue(self.team_page.edit_team_button_present)
+
+        self.verify_form_values_on_team_page(
```
Helper method with lines 914-932?

cahrens commented:
I noted a few minor things about the tests, and the breadcrumbs (and informational banner?) still seem to be in flux. Otherwise :+1: 

Note that I added the i18n issues to https://openedx.atlassian.net/browse/TNL-3125 (I'm still not sure what else that ticket is intended to cover).

muzaffaryousaf commented:
@catong @marcotuts @explorerleslie 
>@marcotuts I discussed this with @explorerleslie, and we think that the informational banner is not really necessary, given that the warning about notifying teams of changes is always visible under the "Edit Team" heading, and also that course staff can see the number of members in a team as they go in to edit it. What would you think about not using the banner at all? If we got rid of the banner, I would tweak the wording under "Edit Team" to use the version we currently use in the banner ("If you make significant changes, make sure you notify members of the team before making these changes.")

I've no issue in removing the message banner and will update the text as suggested but i wish we could have decided this in earlier/design stages that could save a lot of time & effort.

muzaffaryousaf commented:
@explorerleslie @marcotuts @catong @symbolist I've addressed all the comments & suggestions. Would you give it a final look. I'm planning to merge it, this evening. Thanks.

muzaffaryousaf commented:
@cahrens Thanks for the detailed review. I have updated the tests, breadcrumbs and removed the message banner. 
Thanks

catong commented:
@muzaffaryousaf Yes, sorry about the last minute change for the information bar. It was finally seeing everything together that made it clear how repetitive the text was. I took a look at the refreshed sandbox. :+1:

explorerleslie commented:
:+1:

muzaffaryousaf commented:
@catong NP! That change make sense. It's always easy to remove things :)

muzaffaryousaf commented:
@symbolist Waiting for your thumbs up.

symbolist commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -629,78 +629,124 @@ def test_navigation_links(self):
+        self.create_or_edit_team_page.value_for_dropdown_field(field_id='country', value='Pakistan')
 
-    def test_user_can_see_create_team_page(self):
+    def verify_team_data(self):
```
This is about the verifying that the fields are present with the right help messages. Can we make this `verify_all_fields_exist`?

symbolist commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -629,78 +629,124 @@ def test_navigation_links(self):
-        self.create_team_page = CreateTeamPage(self.browser, self.course_id, self.topic)
-        self.team_name = 'Avengers'
+    topic = {'name': 'Example Topic', 'id': 'example_topic', 'description': 'Description'}
+    team_name = 'Avengers'
```
Is a constant. Should be all caps.

symbolist commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -340,10 +384,25 @@
+                    if (options.parentTeam) {
+                        breadcrumbs.push({
+                            title: options.parentTeam.get('name'),
+                            url: 'teams/' + options.topicID + '/' + options.parentTeam.id
```
This should be `#teams`.

symbolist commented:
@muzaffaryousaf I have completed my second pass. I just had a few small suggestions. But otherwise this is good to go. :+1:.

muzaffaryousaf commented:
jenkins run js

muzaffaryousaf commented:
jenkins run lettuce

muzaffaryousaf commented:
jenkins run python

muzaffaryousaf commented:
jenkins run bokchoy
